File: src/ast.rs
//! The typed Markdown AST and the document-level transforms that run on it.
//!
//! Everything the old string pipeline re-parsed with regexes (heading levels,
//! link-only blocks, visible widths, blank runs) is expressed structurally
//! here, so the renderer below can emit output in one deterministic pass.

/// A block-level Markdown node. `Table` rows are already column-pruned; the
/// first row is the header row.
#[derive(Clone, Debug)]
pub enum Block {
    Heading { level: usize, inlines: Vec<Inline> },
    Paragraph { inlines: Vec<Inline> },
    List { ordered: bool, items: Vec<ListItem> },
    Table { rows: Vec<Vec<Vec<Inline>>> },
    Code { body: String },
    Quote { blocks: Vec<Block> },
    Rule,
}

#[derive(Clone, Debug)]
pub struct ListItem {
    /// 1-based position among the source `<li>` children; empty items keep
    /// their number so ordered lists number like the source HTML does.
    pub number: usize,
    pub inlines: Vec<Inline>,
    pub sub_lists: Vec<Block>,
}

/// An inline Markdown node. `Text` holds normalised source text (unicode
/// escapes decoded, whitespace collapsed) that the renderer escapes;
/// `Raw` holds already-rendered Markdown that is emitted verbatim.
#[derive(Clone, Debug)]
pub enum Inline {
    Text(String),
    Raw(String),
    Code(String),
    Strong(Vec<Inline>),
    Emph(Vec<Inline>),
    Link {
        /// Normalised, already-escaped display text.
        text: String,
        href: String,
    },
    LineBreak,
}

/// Drop headings that introduce nothing: a heading whose following block is
/// another heading at the same or shallower level (e.g. an empty `## REVIEWERS`
/// section in a Bitbucket PR sitting right before `## NEW ACTIVITY`). Headings
/// followed by content, by a deeper sub-heading, or at end of document are kept.
pub fn drop_empty_sections(blocks: Vec<Block>) -> Vec<Block> {
    let levels: Vec<Option<usize>> = blocks
        .iter()
        .map(|b| match b {
            Block::Heading { level, .. } => Some(*level),
            _ => None,
        })
        .collect();
    blocks
        .into_iter()
        .enumerate()
        .filter(|(i, _)| match levels[*i] {
            None => true,
            Some(lvl) => !matches!(levels.get(i + 1), Some(Some(next)) if *next <= lvl),
        })
        .map(|(_, b)| b)
        .collect()
}

/// Remap the distinct heading levels actually present to a contiguous `1..n`
/// range. Emails routinely jump levels — a sidebar promo marked up as `<h1>`
/// with the real sections as `<h4>` — which the uniform shift preserves as a
/// jarring `#` → `####`. Rank-mapping renders that as `#` → `##`.
pub fn compress_heading_levels(blocks: Vec<Block>) -> Vec<Block> {
    let mut used: Vec<usize> = blocks
        .iter()
        .filter_map(|b| match b {
            Block::Heading { level, .. } => Some(*level),
            _ => None,
        })
        .collect();
    used.sort_unstable();
    used.dedup();
    if used.len() < 2 {
        return blocks;
    }
    blocks
        .into_iter()
        .map(|b| match b {
            Block::Heading { level, inlines } => {
                let rank = used.iter().position(|&l| l == level).unwrap() + 1;
                Block::Heading {
                    level: rank,
                    inlines,
                }
            }
            other => other,
        })
        .collect()
}

/// Join a run of consecutive blocks that are each a single short link into one
/// ` · `-separated line. Vendor emails stack nav bars, framework-badge rows and
/// footers (`[Next.js]`, `[Nuxt]`, … / `[Unsubscribe]`, `[Privacy notice]`)
/// vertically — one block per link — which renders as a tall column of links.
/// The short-text cap keeps article/post-title link lists (whose text is long)
/// on their own lines.
pub fn join_link_rows(blocks: Vec<Block>) -> Vec<Block> {
    let mut out: Vec<Block> = Vec::new();
    let mut run: Vec<Block> = Vec::new();
    for b in blocks {
        if is_short_link(&b) {
            run.push(b);
        } else {
            flush_run(&mut run, &mut out);
            out.push(b);
        }
    }
    flush_run(&mut run, &mut out);
    out
}

fn flush_run(run: &mut Vec<Block>, out: &mut Vec<Block>) {
    if run.len() < 2 {
        out.append(run);
        return;
    }
    // Vendor nav bars and footers sometimes repeat the same CTA link twice
    // in one row; a repeated (label, href) pair in a joined row is dropped.
    let mut merged: Vec<Inline> = Vec::new();
    let mut seen: Vec<(&str, &str)> = Vec::new();
    for b in run.iter() {
        let Block::Paragraph { inlines } = b else {
            continue;
        };
        for inline in inlines {
            let Inline::Link { text, href } = inline else {
                continue;
            };
            let key = (text.as_str(), href.as_str());
            if seen.contains(&key) {
                continue;
            }
            if !merged.is_empty() {
                merged.push(Inline::Raw(" · ".to_string()));
            }
            seen.push(key);
            merged.push(inline.clone());
        }
    }
    run.clear();
    if !merged.is_empty() {
        out.push(Block::Paragraph { inlines: merged });
    }
}

fn is_short_link(b: &Block) -> bool {
    match b {
        Block::Paragraph { inlines } => match inlines.as_slice() {
            [Inline::Link { text, .. }] => text.chars().count() <= 30,
            _ => false,
        },
        _ => false,
    }
}

File: go/internal/ast/ast.go
// Package ast is the typed Markdown AST and the document-level transforms
// that run on it, mirroring src/ast.rs.
package ast

import (
	"github.com/stubbe/html-to-md/internal/text"
)

type Table = [][][]Inline // rows of cells of inlines

type BlockKind uint8

const (
	KindHeading BlockKind = iota
	KindParagraph
	KindList
	KindTable
	KindCode
	KindQuote
	KindRule
)

type Block struct {
	Kind    BlockKind
	Level   int // heading level
	Inlines []Inline
	Ordered bool
	Items   []ListItem
	Rows    Table
	Body    string // code body
	Blocks  []Block
}

type ListItem struct {
	// Number is the 1-based position among the source <li> children; empty
	// items keep their number so ordered lists number like the source.
	Number   int
	Inlines  []Inline
	SubLists []Block
}

// Inline: Text carries normalised source text (escapes decoded, whitespace
// collapsed) that the renderer escapes; Raw holds already-rendered Markdown
// emitted verbatim.
type Inline struct {
	Kind uint8
	// Text/Raw/Code content (Code and Link text are rendered-verbatim rules)
	S     string
	Inner []Inline
	// Text/Href for links
	LinkText string
	Href     string
}

const (
	IText uint8 = iota
	IRaw
	ICode
	IStrong
	IEmph
	ILink
	ILineBreak
)

func T(s string) Inline { return Inline{Kind: IText, S: s} }
func R(s string) Inline { return Inline{Kind: IRaw, S: s} }
func C(s string) Inline { return Inline{Kind: ICode, S: s} }
func Brk() Inline       { return Inline{Kind: ILineBreak} }

// DropEmptySections drops headings that introduce nothing: a heading whose
// following block is another heading at the same or shallower level (e.g. an
// empty `## REVIEWERS` section right before `## NEW ACTIVITY`). Headings
// followed by content, a deeper sub-heading, or end of document are kept.
func DropEmptySections(blocks []Block) []Block {
	levels := make([]int, len(blocks)) // 0 = not a heading
	for i, b := range blocks {
		if b.Kind == KindHeading {
			levels[i] = b.Level
		} else {
			levels[i] = 0
		}
	}
	out := make([]Block, 0, len(blocks))
	for i, b := range blocks {
		lvl := levels[i]
		if lvl == 0 {
			out = append(out, b)
			continue
		}
		if i+1 < len(blocks) && levels[i+1] != 0 && levels[i+1] <= lvl {
			continue
		}
		out = append(out, b)
	}
	return out
}

// CompressHeadingLevels remaps the distinct heading levels actually present
// to a contiguous 1..n range.
func CompressHeadingLevels(blocks []Block) []Block {
	seen := map[int]bool{}
	var used []int
	for _, b := range blocks {
		if b.Kind == KindHeading && !seen[b.Level] {
			seen[b.Level] = true
			used = append(used, b.Level)
		}
	}
	sortInts(used)
	if len(used) < 2 {
		return blocks
	}
	out := make([]Block, len(blocks))
	for i, b := range blocks {
		if b.Kind == KindHeading {
			b.Level = indexOf(used, b.Level) + 1
		}
		out[i] = b
	}
	return out
}

func sortInts(s []int) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && s[j] < s[j-1]; j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}

func indexOf(s []int, v int) int {
	for i, x := range s {
		if x == v {
			return i
		}
	}
	return -1
}

// JoinLinkRows joins a run of consecutive blocks that are each a single
// short link into one " · "-separated line (vendor nav bars, badge rows,
// footer link columns). The short-text cap keeps article/post-title link
// lists on their own lines.
func JoinLinkRows(blocks []Block) []Block {
	out := make([]Block, 0, len(blocks))
	var run []Block
	flush := func() {
		if len(run) >= 2 {
			if merged := mergeLinkRun(run); merged != nil {
				out = append(out, Block{Kind: KindParagraph, Inlines: merged})
			}
		} else {
			out = append(out, run...)
		}
		run = nil
	}
	for _, b := range blocks {
		if isShortLink(&b) {
			run = append(run, b)
		} else {
			flush()
			out = append(out, b)
		}
	}
	flush()
	return out
}

func isShortLink(b *Block) bool {
	if b.Kind != KindParagraph || len(b.Inlines) != 1 {
		return false
	}
	in := &b.Inlines[0]
	if in.Kind != ILink {
		return false
	}
	// Bound by display width, not raw characters: CJK and emoji occupy
	// twice the terminal columns, so a 30 *code-point* cap would let wide
	// glyphs wrap a joined nav bar.
	return text.Width(in.LinkText) <= 30
}

// mergeLinkRun concatenates the links of the run; a repeated (label, href)
// pair is dropped (vendors repeat the same CTA twice in one row).
func mergeLinkRun(run []Block) []Inline {
	var merged []Inline
	type key struct{ t, h string }
	seen := map[key]bool{}
	for _, b := range run {
		for _, in := range b.Inlines {
			if in.Kind != ILink {
				continue
			}
			k := key{in.LinkText, in.Href}
			if seen[k] {
				continue
			}
			if len(merged) > 0 {
				merged = append(merged, R(" · "))
			}
			seen[k] = true
			merged = append(merged, in)
		}
	}
	if len(merged) == 0 {
		return nil
	}
	return merged
}
